include/deal.II/lac/solver_gmres.h

```diff
@@ -730,7 +731,109 @@ namespace internal
 {
   namespace SolverGMRESImplementation
   {
-    template <class VectorType>
+    template <typename VectorType, typename Enable = void>
+    struct is_dealii_compatible_distributed_vector;
+
+    template <typename VectorType>
+    struct is_dealii_compatible_distributed_vector<
+      VectorType,
+      typename std::enable_if<!internal::is_block_vector<VectorType>>::type>
+    {
+      static constexpr bool value = std::is_same<
+        VectorType,
+        LinearAlgebra::distributed::Vector<typename VectorType::value_type,
+                                           MemorySpace::Host>>::value;
+    };
+
+
+
+    template <typename VectorType>
+    struct is_dealii_compatible_distributed_vector<
+      VectorType,
+      typename std::enable_if<internal::is_block_vector<VectorType>>::type>
+    {
+      static constexpr bool value = std::is_same<
+        typename VectorType::BlockType,
+        LinearAlgebra::distributed::Vector<typename VectorType::value_type,
+                                           MemorySpace::Host>>::value;
+    };
+
+
+
+    template <typename VectorType,
+              std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
+                * = nullptr>
+    unsigned int
+    n_blocks(const VectorType &)
+    {
+      return 1;
+    }
+
+
+
+    template <typename VectorType,
+              std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
+                nullptr>
+    unsigned int
+    n_blocks(const VectorType &vector)
+    {
+      return vector.n_blocks();
+    }
+
+
+
+    template <typename VectorType,
+              std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
+                * = nullptr>
+    VectorType &
+    block(VectorType &vector, const unsigned int b)
+    {
+      AssertDimension(b, 0);
+      (void)b;
+      return vector;
+    }
+
+
+
+    template <typename VectorType,
+              std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
+                * = nullptr>
+    const VectorType &
+    block(const VectorType &vector, const unsigned int b)
+    {
+      AssertDimension(b, 0);
+      (void)b;
+      return vector;
+    }
+
+
+
+    template <typename VectorType,
+              std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
+                nullptr>
+    typename VectorType::BlockType &
+    block(VectorType &vector, const unsigned int b)
+    {
+      return vector.block(b);
+    }
+
+
+
+    template <typename VectorType,
+              std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
+                nullptr>
+    const typename VectorType::BlockType &
+    block(const VectorType &vector, const unsigned int b)
+    {
+      return vector.block(b);
+    }
+
+
+
+    template <class VectorType,
+              std::enable_if_t<
+                !is_dealii_compatible_distributed_vector<VectorType>::value,
+                VectorType> * = nullptr>
     void
     Tvmult_add(const unsigned int dim,
```

Comment:
I realize we did not have those before, but would you mind adding a short comment about what this function does?